### PR TITLE
AVM 7 spec changes

### PR DESCRIPTION
**Pull request by jannotti**
Changes planned for consensus v34 (AVM v7)

### CODE REVIEW

File: dev/TEAL_opcodes.md
Comment (fabrice102) [inline dev/TEAL_opcodes.md:787]: Why do you allow `\r` and `\n`?
Are they allowed anywhere or just at the beginning and end?

File: dev/TEAL.md
Comment (fabrice102) [inline dev/TEAL.md:452]:
```suggestion
| 19 | AssetSender | []byte |      | 32 byte address. Used to move asset from AssetSender if Sender is the Clawback address of the asset. |
```

File: dev/TEAL.md
Comment (fabrice102) [inline dev/TEAL.md:496]: Is it for key registration transaction
```suggestion
| 63 | StateProofPK | []byte | v6  | 64 byte state proof public key |
```
I'm afraid that the word "commitment" can be confusing. The fact that it is a commitment is mostly not relevant to the developer.

File: dev/TEAL.md
Comment (fabrice102) [inline dev/TEAL.md:640]: Is this change retroactive for v6?

File: dev/TEAL_opcodes.md
Comment (fabrice102) [inline dev/TEAL_opcodes.md:391]:
```suggestion
| 19 | AssetSender | []byte |      | 32 byte address. Used to move asset from AssetSender if Sender is the Clawback address of the asset. |
```

File: dev/TEAL_opcodes.md
Comment (fabrice102) [inline dev/TEAL_opcodes.md:435]:
```suggestion
| 63 | StateProofPK | []byte | v6  | 64 byte state proof public key |
```

File: dev/TEAL_opcodes.md
Comment (fabrice102) [inline dev/TEAL_opcodes.md:796]:
```suggestion
WARNING: Usage should be restricted to very rare use cases.
JSON decoding is expensive and is very limited.
In addition, JSON objects are large and not optimized for size.
Almost all smart contracts should be serializing data using simpler-to-deserialize methods (such as using the ABI - https://arc.algorand.foundation/ARCs/arc-0004).
This opcode should only be used in cases where JSON is absolutely the only available option, e.g. when a third-party only accepts to sign specific JSON object.
```

File: dev/TEAL.md
Comment (fabrice102) [inline dev/TEAL.md:337]:
```suggestion
| `base64_decode e` | Decode Base-64 encoded byte-array A. Fail if A is not properly encoded. WARNING: Usage should be restricted to very rare use cases. |
```

File: dev/TEAL.md
Comment (fabrice102) [inline dev/TEAL.md:338]:
```suggestion
| `json_ref r` | Return key B's value from a [valid](jsonspec.md) utf-8 encoded json object string A. Fail if B's value can not be represented by the type R. WARNING: Usage should be restricted to very rare use cases. |
```

File: dev/TEAL_opcodes.md
Comment (fabrice102) [inline dev/TEAL_opcodes.md:786]:
```suggestion
WARNING: Usage should be restricted to very rare use cases.
In almost all cases, smart contracts should directly handle non-encoded byte-strings.
This opcode should only be used in cases where base64 is absolutely the only available option, e.g. when a third-party only accepts to sign specific base64 strings.
```

File: dev/TEAL_opcodes.md
Comment (jannotti) [reply dev/TEAL_opcodes.md:787]: They are allowed anywhere: "`\n` and `\r` are allowed but completely ignored".  It seems to be common to allow such, presumably from legacy uses of b64 like putting them in an email message with limited line length.  This is something the RFC says some implementation do (the Go implementation does), not something we made up out of thin air.

File: dev/TEAL.md
Comment (jannotti) [reply dev/TEAL.md:452]: Agree that it should be reworded to describe the field, rather than sounding like the field is a command.  How about:
"32 byte address. The source of assets if Sender is the Asset's Clawback address."

The change to "Asset's Clawback address" at the end is to match the syntax of "AssetAmount", which reads:
"value in Asset's units"

Maybe "value" that should be capitalized for AssetAmount.

File: dev/TEAL.md
Comment (jannotti) [reply dev/TEAL.md:640]: yes

File: dev/TEAL.md
Comment (jannotti) [reply dev/TEAL.md:337]: I'm very happy with this warning on `json_ref` but I think we can be a bit softer here, since at least b64 is well defined and not all that slow.  How about: "Users of `base64_decode` should consider whether their input could be supplied unencoded to save transaction space and computation within the AVM"

File: dev/TEAL_opcodes.md
Comment (shaih) [inline dev/TEAL_opcodes.md:80]: How is the curve specified? What happens if a ve5/v6 program tries to

Comment (jannotti) [reply dev/TEAL_opcodes.md:80]: The curve is given as an immediate argument to `ecdsa_pk_decompress`, so the TEAL reads:
```
ecdsa_pk_decompress Secp256k1
```
or
```
ecdsa_pk_decompress Secp256r1
```
However, the use of such immediates is version-gated, as described here:
```
| Index | Name | In | Notes |
| - | ------ | - | --------- |
| 0 | Secp256k1 |      | secp256k1 curve, used in Bitcoin |
| 1 | Secp256r1 | v7  | secp256r1 curve, NIST standard |
```

Since Secp256r1 is v7, it will neither assemble that TEAL, not execute the corresponding bytecode (it would fail at runtime if you put the bytecode together "by hand" in a v5 program).

File: dev/TEAL.md
Comment (fabrice102) [reply dev/TEAL.md:451]: "32 byte address. The source of assets if Sender is the Asset's Clawback address."

looks good

File: dev/TEAL.md
Comment (fabrice102) [reply dev/TEAL.md]: Indeed, "Users of base64_decode should consider whether their input could be supplied unencoded to save transaction space and computation within the AVM" looks good to me

File: dev/TEAL.md
Comment (fabrice102) [inline dev/TEAL.md:760]: Wasn't this constraint removed?

File: dev/TEAL_opcodes.md
Comment (fabrice102) [inline dev/TEAL_opcodes.md:1404]: The official name is:
ECVRF-EDWARDS25519-SHA512-Elligator2 from https://tools.ietf.org/id/draft-irtf-cfrg-vrf-06.html

**Review by fabrice102 — CHANGES_REQUESTED** [review]

**Comment by algorandskiy** [comment]
Please merge this after #70 or #67 since we'll be making two protocol upgrades within a single release (see https://github.com/algorand/go-algorand/issues/4232) and AVM changes must go after enabling 1001 rounds back lookback on all nodes.

File: dev/TEAL_opcodes.md
Comment (jannotti) [reply dev/TEAL_opcodes.md:1404]: We don't want to make that the constant name, do we?

Also, I thought someone said that the ietf draft ended up being slightly different somehow?

File: dev/TEAL.md
Comment (jannotti) [reply dev/TEAL.md:760]: Correct, I missed this reference. I'll change.

File: dev/TEAL.md
Comment (fabrice102) [inline dev/TEAL.md:748]:
```suggestion
in the transaction group. 

Concretely, the above requirement is translated as follows:
A v1 program included in a transaction group
```

Comment (jannotti) [reply dev/TEAL.md]: I've committed that. If this seems good, please approve but hold off on merging, as per @algorandskiy

Comment (algorandskiy) [reply dev/TEAL.md]: I withdraw my prev request to hold the merge - I've used the block size increase commit hash for v33 and any other final merge can be used for v34. Please go ahead and merge if it looks good.

File: dev/TEAL_opcodes.md
Comment (fabrice102) [inline dev/TEAL_opcodes.md:1405]:
```suggestion
`VrfAlgorand` is the VRF used in Algorand, that is the ECVRF-ED25519-SHA512-Elligator2 VRF specified in the IETF internet draft draft-irtf-cfrg-vrf-03 (https://datatracker.ietf.org/doc/draft-irtf-cfrg-vrf/03/).
```

**Review by fabrice102 — APPROVED** [review]

**Pull request merged**
